### ISSUE TITLE
Segment fault when destroy multi instance at multi thread

### ISSUE THREAD

**Issue by aggresss** [eclipse-paho/paho.mqtt.c#700]
Segment fault when destroy multi instance at multi thread

```shell
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Core was generated by `./tests/unit/unit_test_mqtt --instance 5 --loop 5'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x0000560e95ba0ce2 in ListNextElement ()
[Current thread is 1 (Thread 0x7fcdc16b7700 (LWP 376))]
(gdb) bt -5
#0  0x0000560e95ba0ce2 in ListNextElement ()
#1  0x0000560e95bac2ce in MQTTAsync_checkTimeouts ()
#2  0x0000560e95bac3b1 in MQTTAsync_sendThread ()
#3  0x00007fcdc42b56db in start_thread (arg=0x7fcdc16b7700) at pthread_create.c:463
#4  0x00007fcdc3fde88f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

**Comment by icraggs** [comment]
You must make sure the client is fully disconnected before calling destroy, otherwise data structures will be removed which are still in use.  Multi-threaded examples of use are shown within the test suites.  

I see you have a test program.  Attaching that here might be useful.

**Comment by icraggs** [comment]
Fixed by #707?
